llama_cpp/llama_cache.py

```diff
@@ -13,38 +14,93 @@
 
 from .llama_types import *
 
+@dataclass(eq=True, frozen=True)
+class LlamaCacheKey:
+    """A key in a LlamaCache. Stores tokens to key by. Also stores
+    information about active LoRA adapters, because we need different
+    cached values for different active adapters, even for the same tokens."""
+    active_lora_adapters: Tuple[Tuple[str, float], ...]
+    tokens: Tuple[int, ...]
+
+    def __post_init__(self):
+        if not isinstance(self.tokens, tuple):
+            raise ValueError("tokens must be a tuple")
 
 class BaseLlamaCache(ABC):
     """Base cache class for a llama.cpp model."""
 
     def __init__(self, capacity_bytes: int = (2 << 30)):
         self.capacity_bytes = capacity_bytes
 
+    def _convert_to_cache_key(self, key: Union[Sequence[int], LlamaCacheKey]) -> LlamaCacheKey:
+        """Convert raw tokens to a key if needed"""
+        if type(key) == LlamaCacheKey:
+            return key
+        else:
+            return LlamaCacheKey(active_lora_adapters=(), tokens=tuple(key))
```

Comment:
Provides backwards compatibility for existing code which passes a `Sequence[int]` to access the cache. This format is still supported, although existing cached values won't be found.